File: lang-core/src/main/java/com/jaitlapps/differ/reader/impl/StringReader.java
package com.jaitlapps.differ.reader.impl;

import com.google.common.collect.ImmutableSet;
import com.jaitlapps.differ.reader.Reader;

import java.util.Set;

public class StringReader implements Reader {

    private static final Set<Character> SEPARATORS = ImmutableSet.<Character>builder()
            .add('+')
            .add('-')
            .add('*')
            .add('/')
            .add('=')
            .add('[')
            .add(']')
            .add('^')
            .add(' ')
            .add('\n')
            .add('\r')
            .add(',')
            .build();

    private final String stream;
    private final char[] charStream;
    private final int lengthStream;

    private int startPosition;
    private int endPosition;

    public StringReader(String stream) {
        this.stream = stream;
        this.charStream = stream.toCharArray();
        this.lengthStream = stream.length();

        this.startPosition = 0;
        this.endPosition = 0;
    }

    @Override
    public String readNextWord() {
        while (endPosition < lengthStream && charStream[endPosition] == ' ') {
            endPosition++;
        }

        startPosition = endPosition;

        if (endPosition < lengthStream) {
            if (endPosition < lengthStream && SEPARATORS.contains(charStream[endPosition])) {
                endPosition++;
            } else {
                while (endPosition < lengthStream && charStream[endPosition] != ' ') {
                    if (SEPARATORS.contains(charStream[endPosition])) {
                        break;
                    }
                    endPosition++;
                }
            }

            return stream.substring(startPosition, endPosition);
        }

        return null;
    }

    public int getStartPosition() {
        return startPosition;
    }

    public int getEndPosition() {
        return endPosition;
    }
}

File: lang-core/src/main/kotlin/com/jaitlapps/differ/reader/impl/StringReader.kt
package com.jaitlapps.differ.reader.impl

import com.google.common.collect.ImmutableSet
import com.jaitlapps.differ.model.Word
import com.jaitlapps.differ.reader.Reader

class StringReader(private val stream: String) : Reader {
    private val charStream: CharArray
    private val lengthStream: Int

    private var startPosition: Int = 0;
    private var endPosition: Int = 0;

    init {
        this.charStream = stream.toCharArray()
        this.lengthStream = stream.length
    }

    override fun readNextWord(): Word? {

        while (endPosition < lengthStream && charStream[endPosition] == ' ') {
            endPosition++
        }

        startPosition = endPosition

        if (endPosition < lengthStream) {
            if (endPosition < lengthStream && SEPARATORS.contains(charStream[endPosition])) {
                endPosition++
            } else {
                while (endPosition < lengthStream && charStream[endPosition] != ' ') {
                    if (SEPARATORS.contains(charStream[endPosition])) {
                        break
                    }
                    endPosition++
                }
            }

            return Word(startPosition, endPosition, stream.substring(startPosition, endPosition))
        }

        return null
    }

    companion object {
        private val SEPARATORS = ImmutableSet.builder<Char>().add('+').add('-').add('*').add('/').add('=').add('[').add(']').add('^').add(' ').add('\n').add('\r').add(',').build()
    }
}
